File: dynamic_programming/top_down_matrix_chain_multiplication.py
import sys


def matrix_chain_order(p, idx_of_start, idx_of_end):
    results = [
        [sys.maxsize for _ in range(idx_of_end + 1)]
        for _ in range(idx_of_end + 1)
    ]

    return matrix_chain_order_internal(p, idx_of_start, idx_of_end, results)


def matrix_chain_order_internal(p, idx_of_start, idx_of_end, results):
    if results[idx_of_start][idx_of_end] < sys.maxsize:
        return results[idx_of_start][idx_of_end]

    result = 0
    if idx_of_start < idx_of_end:
        result = sys.maxsize
        for k in range(idx_of_start, idx_of_end):
            result = min(result, matrix_chain_order_internal(p, idx_of_start, k, results)
                         + matrix_chain_order_internal(p, k + 1, idx_of_end, results)
                         + p[idx_of_start] * p[k + 1] * p[idx_of_end + 1])

    results[idx_of_start][idx_of_end] = result

    return result


if __name__ == '__main__':
    p = [30, 35, 15, 5, 10, 20, 25]

    print("  i |", end='')
    for i in range(len(p) - 1):
        print(f"{i:^7}|", end='')
    print('')
    print("A[i]|", end='')
    for i in range(len(p) - 1):
        print(f"{p[i]:>3}*{p[i + 1]:>3}|", end='')
    print('')

    diff = 0
    while diff < len(p) - 1:
        print(f"--- diff = {diff} ---")

        i, j = 0, diff
        while j < len(p) - 1:
            result = matrix_chain_order(p, i, j)

            print(f"i = {i}, j = {j} / result = {result}")

            i += 1
            j += 1

        diff += 1

File: dynamic_programming/top_down_matrix_chain_multiplication.go
//go:build ignore

package main

import (
	"fmt"
	"math"
)

func matrixChainOrder(p []int, i, j int) int {
	if i == j {
		return 0
	}

	result := math.MaxInt
	for k := i; k < j; k++ {
		result = min(result, matrixChainOrder(p, i, k)+matrixChainOrder(p, k+1, j)+p[i]*p[k+1]*p[j+1])
	}

	return result
}

func main() {
	p := []int{30, 35, 15, 5, 10, 20, 25}

	for diff := 0; diff < len(p)-1; diff = diff + 1 {
		fmt.Printf("--- diff = %d ---\n", diff)

		i := 0
		for j := i + diff; j < len(p)-1; j += 1 {
			result := matrixChainOrder(p, i, j)

			fmt.Printf("(i, j) = (%d, %d) / result = %d\n", i, j, result)

			i += 1
		}
	}
}
